### ISSUE TITLE
Initializer Performance Brickwall

### ISSUE THREAD

**Issue by Tshimanga** [chapel-lang/chapel#9167]
In `NumSuch` and `Chingon` we have the following classes `NamedMatrix` and `Graph` defined in the following way

````
class NamedMatrix {
  var D: domain(2),
      SD = CSRDomain(D),
      X: [SD] real,  // the actual data
      rows: BiMap = new BiMap(),
      cols: BiMap = new BiMap();
````
and
````
class Graph: NamedMatrix {
  var name: string,
      directed: bool = false,
      bipartite: bool = false,
      verts: BiMap = new BiMap(),
      uVerts: BiMap = new BiMap(),
      vVerts: BiMap = new BiMap();
````

As shown above, `NamedMatrix` and `Graph` merely take regular matrices and imbue them with additional structure. Understood in this way, one ought to be able to take any regular matrix and then "dress it" in the clothing of a `NamedMatrix`. Similarly given a `NamedMatrix` one ought to be able to dress that in the clothing of a `Graph`.

Typically this would be done by defining a class constructor that, for instance, takes in a `NamedMatrix` 
and builds a `Graph` out of it like in the following
````
  proc init(N: NamedMatrix) {
    super.init(N);
    this.verts = this.rows.uni(this.cols);
    this.uVerts = this.rows;
    this.vVerts = this.cols;
  }
````

The above constructor *has* been demonstrated to work properly; however, when used with a `NamedMatrix` of size `{1..140,000, 1..140,000}` with ~10^6 non-zero entries, it takes ~30 minutes to construct the object. This is unreasonably large considering that the `NamedMatrix` took only ~60s to build element by element from postgres.

As a Chapel user, I'd like to be able to use my constructors *at scale*. How do I do this?

**Comment by lydia-duncan** [comment]
I'd be curious to see if you observe similar behavior with old style constructors, or if the problem is limited to new style initializers

**Comment by mppf** [comment]
I'm supposing that the copy / assignment in the initializer is the source of this speed issue. That may or may not be correct and would be a good 1st thing to establish.

I think initializers should generally take their arguments by 'in' intent and that the initializers implementation needs to be updated to avoid a copy when initializing a field from such an argument. I think @benharsh is looking at this at some not-to-distant point.

When you create your Graph, what is the argument to the constructor? A variable or the result of a function call? (If it's the result of a function call, we should be able to entirely remove the copy that I'm supposing is causing the speed issue).

Of course the alternative approach is to investigate why that copy is so slow.

**Comment by bradcray** [comment]
While looking through a related example yesterday, it looked to me like the sparse domain assignment operator was implemented in a very general and expensive manner that could potentially be made much cheaper in the initialization context (because it could assume it's starting from scratch rather than updating a LHS domain with potential arrays hanging off of it).  I'm not sure whether that same thing could be a factor here or not.  What is BiMap?  What does NamedMatrix.init() look like?

**Comment by Tshimanga** [comment]
@mppf but why would creating a duplicate of an object that took ~60s to build take 30minutes.

I'm not familiar with 'in' intent. Can you point me to an example/doc?

The `Graph` is built with
````
var nm = NamedMatrixFromPGSquare(con, edgeTable, fromField, toField, wField = "NONE"); // takes 60s
var g = new Graph(nm); // takes ~30minutes
````

**Comment by Tshimanga** [comment]
@bradcray `BiMap` is an isomorphism between two indexing sets
````
  class BiMap {
    var keys: domain(string),
        ids:  [keys] int,
        idxkey: domain(int),
        idx: [idxkey] string;
````
The `NamedMatrix.init()`s are
````
   proc init() {
     this.complete();
   }

   proc init(X) {
     this.D = {X.domain.dim(1), X.domain.dim(2)};
     this.SD = X.domain;
     this.complete();
     this.loadX(X);
   }

   proc init(X, names: [] string) {
     this.init(X);
     try! this.setRowNames(names);
     try! this.setColNames(names);
   }

   proc init(X, rownames: [] string, colnames: [] string) {
     this.init(X);
     try! this.setRowNames(rownames);
     try! this.setColNames(colnames);
   }

   proc init(rownames: [] string, colnames: [] string) {
     this.D = {1..rownames.size, 1..colnames.size};
     this.complete();
     try! this.setRowNames(rownames);
     try! this.setColNames(colnames);
   }

   proc init(N: NamedMatrix) {
     this.init(N.X);
     this.rows = N.rows;
     this.cols = N.cols;
   }

   proc init(X, rows: BiMap, cols: BiMap) {
     this.init(X);
     this.rows = rows;
     this.cols = cols;
   }
````

**Comment by mppf** [comment]
If your initializer used the 'in' intent, and 'in' intent initializers are improved,
```
var g = new Graph(NamedMatrixFromPGSquare(con, edgeTable, fromField, toField, wField = "NONE"));
```
would remove the copy during initialization.

https://chapel-lang.org/docs/master/primers/procedures.html#argument-intents has info on in intents, and they're described in more detail (somewhere) in the language specification. The copy avoidance I'm talking about is in [CHIP 13](https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/chips/13.rst).

**Comment by Tshimanga** [comment]
@lydia-duncan yes, if I recall correctly. The Initializer Performance Brickwall was present in the old initalizer scheme as well but at the time we had other to-dos.

**Comment by Tshimanga** [comment]
Just so the note doesn't confusion anyone. The merged pull request above doesn't solve the problem haha

**Comment by cassella** [comment]
Can you identify where most of the time is being spent?
